src/core/modules/core/core_cache.cpp

```diff
@@ -0,0 +1,326 @@
+	set_deleter(fdel);
+
+	m_szDocString = doc;
+	m_bUnbound = unbound;
+
+	m_args = args;
+	m_kwargs = kwargs;
+}
+
+
+object CCachedProperty::_callable_check(object function, const char *szName)
+{
+	if (!function.is_none() && !PyCallable_Check(function.ptr()))
+		BOOST_RAISE_EXCEPTION(
+			PyExc_TypeError,
+			"The given %s function is not callable.", szName
+		);
+
+	return function;
+}
+
+object CCachedProperty::_prepare_value(object value)
+{
+	if (!PyGen_Check(value.ptr()))
+		return value;
+
+	if (getattr(value, "gi_frame").is_none())
+		BOOST_RAISE_EXCEPTION(
+			PyExc_ValueError,
+			"The given generator is exhausted."
+		);
+
+	list values;
+	while (true)
+	{
+		try
+		{
+			values.append(value.attr("__next__")());
+		}
+		catch(...)
+		{
+			if (!PyErr_ExceptionMatches(PyExc_StopIteration))
+				throw_error_already_set();
+
+			PyErr_Clear();
+			break;
+		}
+	}
+
+	return values;
+}
+
+void CCachedProperty::_invalidate_cache(PyObject *pRef)
+{
+	try
+	{
+		m_cache[handle<>(pRef)].del();
+	}
+	catch (...)
+	{
+		if (!PyErr_ExceptionMatches(PyExc_KeyError))
+			throw_error_already_set();
+
+		PyErr_Clear();
+	}
+}
+
+void CCachedProperty::_update_cache(object instance, object value)
+{
+	if (m_bUnbound)
+		m_cache[handle<>(
+			PyWeakref_NewRef(
+				instance.ptr(),
+				make_function(
+					boost::bind(&CCachedProperty::_invalidate_cache, this, _1),
+					default_call_policies(),
+					boost::mpl::vector2<void, PyObject *>()
+				).ptr()
+			)
+		)] = value;
+	else
+	{
+		dict cache = extract<dict>(instance.attr("__dict__"));
+		cache[m_name] = value;
+	}
+}
+
+void CCachedProperty::_delete_cache(object instance)
+{
+	try
+	{
+		if (m_bUnbound)
+			m_cache[
+				handle<>(
+					PyWeakref_NewRef(instance.ptr(), NULL)
+				)
+			].del();
+		else
+		{
+			dict cache = extract<dict>(instance.attr("__dict__"));
+			cache[m_name].del();
+		}
+	}
+	catch (...)
+	{
+		if (!PyErr_ExceptionMatches(PyExc_KeyError))
+			throw_error_already_set();
+
+		PyErr_Clear();
+	}
+}
+
+
+object CCachedProperty::get_getter()
+{
+	return m_fget;
+}
+
+object CCachedProperty::set_getter(object fget)
+{
+	m_fget = _callable_check(fget, "getter");
+	return fget;
+}
+
+
+object CCachedProperty::get_setter()
+{
+	return m_fset;
+}
+
+object CCachedProperty::set_setter(object fset)
+{
+	m_fset = _callable_check(fset, "setter");
+	return fset;
+}
+
+
+object CCachedProperty::get_deleter()
+{
+	return m_fdel;
+}
+
+object CCachedProperty::set_deleter(object fdel)
+{
+	m_fdel = _callable_check(fdel, "deleter");
+	return fdel;
+}
+
+
+str CCachedProperty::get_name()
+{
+	return m_name;
+}
+
+object CCachedProperty::get_owner()
+{
+	return m_owner();
+}
+
+
+void CCachedProperty::__set_name__(object owner, str name)
+{
+	m_name = name;
+	m_owner = object(handle<>(PyWeakref_NewRef(owner.ptr(), NULL)));
+}
+
+object CCachedProperty::__get__(object self, object instance, object owner=object())
+{
+	CCachedProperty &pSelf = extract<CCachedProperty &>(self);
```

Comment:
Should we move this line a little bit down (behind the check), since ``pSelf`` isn't used if ``instance`` is ``None``? Or is it done here to let ``extract`` do some validations?

Reply:
No specific reason why it is done there, and I agree, it should be moved and casted only when needed.